alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+    -------
+    The loaded embedding.
+    """
+    src = cfg['src']
+    layers = cfg['layers']
+    typ = cfg['type']
+    if backend == 'tensorflow':
+        emb = load_embedding_tf(src, embedding_type=typ, layers=layers)
+    else:
+        emb = None
+    return emb
+
+
+def _load_tokenizer(cfg: dict) -> AutoTokenizer:
+    """
+    Loads a text tokenizer from a tokenizer config dict.
+
+    Parameters
+    ----------
+    cfg
+        A tokenizer config dict. (see the pydantic schemas).
+
+    Returns
+    -------
+    The loaded tokenizer.
+    """
+    src = cfg['src']
+    kwargs = cfg['kwargs']
+    src = Path(src)
+    tokenizer = AutoTokenizer.from_pretrained(src, **kwargs)
+    return tokenizer
+
+
+def _load_optimizer(cfg: dict,
+                    backend: str) -> Union['tf.keras.optimizers.Optimizer', Callable]:
+    """
+    Loads an optimzier from an optimizer config dict. When backend='tensorflow', the config dict should be in
+    the format given by tf.keras.optimizers.serialize().
+
+    Parameters
+    ----------
+    cfg
+        The optimizer config dict.
+    backend
+        The backend.
+
+    Returns
+    -------
+    The loaded optimizer.
+    """
+    if backend == 'tensorflow':
+        optimizer = load_optimizer_tf(cfg)
+    else:
+        raise NotImplementedError('Loading of pytorch optimizers not currently supported')
+    return optimizer
+
+
+def _get_nested_value(dic: dict, keys: list) -> Any:
+    """
+    Get a value from a nested dictionary.
+
+    Parameters
+    ----------
+    dic
+        The dictionary.
+    keys
+        List of keys to "walk" to nested value.
+        For example, to extract the value `dic['key1']['key2']['key3']`, set `keys = ['key1', 'key2', 'key3']`.
+
+    Returns
+    -------
+    The nested value specified by `keys`.
+    """
+    for key in keys:
+        try:
+            dic = dic[key]
+        except (TypeError, KeyError):
+            return None
+    return dic
+
+
+def _set_nested_value(dic: dict, keys: list, value: Any):
+    """
+    Set a value in a nested dictionary.
+
+    Parameters
+    ----------
+    dic
+        The dictionary.
+    keys
+        List of keys to "walk" to nested value.
+        For example, to set the value `dic['key1']['key2']['key3']`, set `keys = ['key1', 'key2', 'key3']`.
+    value
+        The value to set.
+    """
+    for key in keys[:-1]:
+        dic = dic.setdefault(key, {})
+    dic[keys[-1]] = value
+
+
+def read_config(filepath: Union[os.PathLike, str]) -> dict:
+    """
+    This function reads a detector toml config file and returns a dict specifying the detector.
+
+    Parameters
+    ----------
+    filepath
+        The filepath to the config.toml file.
+
+    Returns
+    -------
+        Parsed toml dictionary.
+    """
+    filepath = Path(filepath)
+
+    cfg = dict(toml.load(filepath))  # toml.load types return as MutableMapping, force to dict
+    logger.info('Loaded config file from %s', str(filepath))
```

Comment:
I knew that logging was an exception to f-strings due to performance overhead :(

Reply:
I've tried to replace all string interpolations. We should either have f-strings for error raising and `.format()`'s for logging.